File: native/keyholder.swift
// keyholder (macOS) — holds key combos down until told to release.
//
// Protocol, one command per line on stdin:
//   D <id> <mods> <key>   press and HOLD under <id>. Either combo field may be "-".
//   U <id>                release the hold owned by <id>
//   T <mods> <key>        TAP a combo without disturbing anything held
//
// Holds are keyed by id because several Stream Deck buttons or pedals can be down at
// once, and each must release only its own keys. A shared key or modifier stays down
// until the last hold that wants it lets go.
//
// Modifier events are posted separately so left/right identity is preserved. The regular
// key event also carries the accumulated modifier flags.
//
// Events are built on a .hidSystemState source and posted to .cghidEventTap, the lowest
// injection point CoreGraphics exposes.
//
// State lives in static storage, not top-level vars: in a single-file Swift program the
// top level is an implicit main(), so a signal handler touching top-level vars would be
// "capturing context" and could not become a C function pointer.

import CoreGraphics
import Foundation

enum Keyholder {
    static let source = CGEventSource(stateID: .hidSystemState)

    /// macOS virtual keycodes. F13-F19 are the useful ones: macOS ignores them, so they
    /// collide with nothing.
    static let keys: [String: CGKeyCode] = [
        "a": 0, "b": 11, "c": 8, "d": 2, "e": 14, "f": 3, "g": 5, "h": 4, "i": 34,
        "j": 38, "k": 40, "l": 37, "m": 46, "n": 45, "o": 31, "p": 35, "q": 12,
        "r": 15, "s": 1, "t": 17, "u": 32, "v": 9, "w": 13, "x": 7, "y": 16, "z": 6,
        "space": 49,
        "0": 29, "1": 18, "2": 19, "3": 20, "4": 21,
        "5": 23, "6": 22, "7": 26, "8": 28, "9": 25,
        "enter": 36, "numpadenter": 76, "tab": 48, "escape": 53, "backspace": 51,
        "delete": 117,
        "insert": 114, "home": 115, "end": 119, "pageup": 116, "pagedown": 121,
        "left": 123, "right": 124, "down": 125, "up": 126,
        "semicolon": 41, "equal": 24, "comma": 43, "minus": 27, "period": 47,
        "slash": 44, "backquote": 50, "leftbracket": 33, "backslash": 42,
        "rightbracket": 30, "quote": 39,
        "f1": 122, "f2": 120, "f3": 99, "f4": 118, "f5": 96, "f6": 97,
        "f7": 98, "f8": 100, "f9": 101, "f10": 109, "f11": 103, "f12": 111,
        "f13": 105, "f14": 107, "f15": 113, "f16": 106, "f17": 64, "f18": 79, "f19": 80,
        "f20": 90,
    ]

    static let modifierKeys: [String: (CGKeyCode, CGEventFlags)] = [
        "ctrl": (59, .maskControl), "lctrl": (59, .maskControl), "rctrl": (62, .maskControl),
        "alt": (58, .maskAlternate), "lalt": (58, .maskAlternate), "ralt": (61, .maskAlternate),
        "shift": (56, .maskShift), "lshift": (56, .maskShift), "rshift": (60, .maskShift),
        "cmd": (55, .maskCommand), "lcmd": (55, .maskCommand), "rcmd": (54, .maskCommand),
    ]

    struct Hold {
        let id: String
        let key: CGKeyCode?
        let modifiers: [(CGKeyCode, CGEventFlags)]
    }

    /// Everything currently held, in press order, so we can always release it — even if
    /// the plugin dies mid-hold. A stuck key is the worst failure here: it would make the
    /// machine unusable until logout.
    static var holds: [Hold] = []

    static func currentFlags() -> CGEventFlags {
        var flags: CGEventFlags = []
        for hold in holds {
            for (_, flag) in hold.modifiers {
                flags.insert(flag)
            }
        }
        return flags
    }

    static func isKeyDown(_ key: CGKeyCode) -> Bool {
        holds.contains { $0.key == key }
    }

    static func isModifierDown(_ modifierKey: CGKeyCode) -> Bool {
        holds.contains { hold in hold.modifiers.contains { $0.0 == modifierKey } }
    }

    static func post(_ key: CGKeyCode, _ flags: CGEventFlags, down: Bool) {
        guard let event = CGEvent(keyboardEventSource: source, virtualKey: key, keyDown: down)
        else { return }
        event.flags = flags
        event.post(tap: .cghidEventTap)
    }

    static func postModifier(_ key: CGKeyCode, _ flags: CGEventFlags) {
        guard let event = CGEvent(keyboardEventSource: source, virtualKey: key, keyDown: true)
        else { return }
        event.type = .flagsChanged
        event.flags = flags
        event.post(tap: .cghidEventTap)
    }

    /// Keys another hold already has down are not pressed again — a second press would be
    /// harmless, but the matching release would not: it would cancel the other hold.
    static func press(id: String, key: CGKeyCode?, modifiers: [(CGKeyCode, CGEventFlags)]) {
        release(id: id)
        var flags = currentFlags()
        for (modifierKey, flag) in modifiers {
            let alreadyDown = isModifierDown(modifierKey)
            flags.insert(flag)
            if !alreadyDown {
                postModifier(modifierKey, flags)
            }
        }
        if let key, !isKeyDown(key) {
            post(key, flags, down: true)
        }
        holds.append(Hold(id: id, key: key, modifiers: modifiers))
    }

    static func release(id: String) {
        guard let index = holds.firstIndex(where: { $0.id == id }) else { return }
        let hold = holds.remove(at: index)

        // The key-up carries the flags that were in effect while it was down, including
        // this hold's own modifiers — those come off afterwards.
        var flags = currentFlags()
        for (_, flag) in hold.modifiers {
            flags.insert(flag)
        }
        if let key = hold.key, !isKeyDown(key) {
            post(key, flags, down: false)
        }

        var remaining = hold.modifiers
        while let (modifierKey, _) = remaining.popLast() {
            if isModifierDown(modifierKey) { continue }
            flags = currentFlags()
            for (_, remainingFlag) in remaining {
                flags.insert(remainingFlag)
            }
            postModifier(modifierKey, flags)
        }
    }

    static func releaseAll() {
        for id in holds.map(\.id).reversed() {
            release(id: id)
        }
    }

    /// Tap a combo *without* disturbing any hold — this is what makes a "before release"
    /// hotkey different from an "after release" one. Modifiers already held are reused
    /// rather than re-posted, and a tap of a key that is currently held is skipped: its
    /// key-up would cancel the hold that owns it.
    static func tapOver(_ key: CGKeyCode?, _ modifiers: [(CGKeyCode, CGEventFlags)]) {
        let heldFlags = currentFlags()
        var extra = modifiers.filter { !isModifierDown($0.0) }

        var flags = heldFlags
        for (modifierKey, flag) in extra {
            flags.insert(flag)
            postModifier(modifierKey, flags)
        }

        if let key {
            if isKeyDown(key) {
                FileHandle.standardError.write(
                    Data("keyholder: skipping tap of the key already held\n".utf8))
            } else {
                post(key, flags, down: true)
                post(key, flags, down: false)
            }
        }

        while let (modifierKey, _) = extra.popLast() {
            flags = heldFlags
            for (_, remainingFlag) in extra {
                flags.insert(remainingFlag)
            }
            postModifier(modifierKey, flags)
        }
    }
}

func onSignal(_ signum: Int32) {
    Keyholder.releaseAll()
    exit(0)
}

setvbuf(stdout, nil, _IONBF, 0)

guard Keyholder.source != nil else {
    FileHandle.standardError.write(Data("keyholder: could not create event source\n".utf8))
    exit(1)
}

for sig in [SIGTERM, SIGINT, SIGHUP] {
    signal(sig, onSignal)
}

func badCommand(_ line: String) {
    FileHandle.standardError.write(Data("keyholder: bad command: \(line)\n".utf8))
}

while let line = readLine(strippingNewline: true) {
    let parts = line.split(separator: " ")
    guard let verb = parts.first else { continue }

    switch verb {
    case "U":
        guard parts.count == 2 else {
            badCommand(line)
            continue
        }
        Keyholder.release(id: String(parts[1]))

    case "D", "T":
        let isHold = verb == "D"
        guard parts.count == (isHold ? 4 : 3) else {
            badCommand(line)
            continue
        }
        let id = isHold ? String(parts[1]) : ""
        let modField = parts[isHold ? 2 : 1]
        let keyName = String(parts[isHold ? 3 : 2]).lowercased()

        let key = keyName == "-" ? nil : Keyholder.keys[keyName]
        if keyName != "-" && key == nil {
            badCommand(line)
            continue
        }
        var modifiers: [(CGKeyCode, CGEventFlags)] = []
        if modField != "-" {
            for name in modField.split(separator: ",") {
                guard let modifier = Keyholder.modifierKeys[String(name).lowercased()] else { continue }
                modifiers.append(modifier)
            }
        }
        guard key != nil || !modifiers.isEmpty else { continue }

        if isHold {
            Keyholder.press(id: id, key: key, modifiers: modifiers)
        } else {
            Keyholder.tapOver(key, modifiers)
        }

    default:
        continue
    }
}

// stdin closed — Stream Deck killed the plugin. Never leave a key down.
Keyholder.releaseAll()

File: native/keyholder.cpp
// keyholder (Windows) — holds key combos down until told to release.
//
// Protocol, one command per line on stdin:
//   D <id> <mods> <key>   press and HOLD under <id>. Either combo field may be "-".
//   U <id>                release the hold owned by <id>
//   T <mods> <key>        TAP a combo without disturbing anything held
//
// Holds are keyed by id because several Stream Deck buttons or pedals can be down at
// once, and each must release only its own keys. A shared key or modifier stays down
// until the last hold that wants it lets go.
//
// Unlike macOS (where a combo is one event carrying modifier flags), SendInput has no
// flags field: the modifier keys must be physically pressed, then the key, and released
// in reverse order. That asymmetry is why the helper owns combo semantics, not the plugin.

#include <windows.h>

#include <algorithm>
#include <iostream>
#include <map>
#include <sstream>
#include <string>
#include <vector>

namespace {

const std::map<std::string, WORD> kKeys = {
    {"a", 'A'}, {"b", 'B'}, {"c", 'C'}, {"d", 'D'}, {"e", 'E'}, {"f", 'F'},
    {"g", 'G'}, {"h", 'H'}, {"i", 'I'}, {"j", 'J'}, {"k", 'K'}, {"l", 'L'},
    {"m", 'M'}, {"n", 'N'}, {"o", 'O'}, {"p", 'P'}, {"q", 'Q'}, {"r", 'R'},
    {"s", 'S'}, {"t", 'T'}, {"u", 'U'}, {"v", 'V'}, {"w", 'W'}, {"x", 'X'},
    {"y", 'Y'}, {"z", 'Z'},
    {"space", VK_SPACE},
    {"0", '0'}, {"1", '1'}, {"2", '2'}, {"3", '3'}, {"4", '4'},
    {"5", '5'}, {"6", '6'}, {"7", '7'}, {"8", '8'}, {"9", '9'},
    {"enter", VK_RETURN}, {"numpadenter", VK_RETURN},
    {"tab", VK_TAB}, {"escape", VK_ESCAPE},
    {"backspace", VK_BACK}, {"delete", VK_DELETE}, {"insert", VK_INSERT},
    {"home", VK_HOME}, {"end", VK_END}, {"pageup", VK_PRIOR},
    {"pagedown", VK_NEXT}, {"left", VK_LEFT}, {"right", VK_RIGHT},
    {"up", VK_UP}, {"down", VK_DOWN},
    {"semicolon", VK_OEM_1}, {"equal", VK_OEM_PLUS}, {"comma", VK_OEM_COMMA},
    {"minus", VK_OEM_MINUS}, {"period", VK_OEM_PERIOD}, {"slash", VK_OEM_2},
    {"backquote", VK_OEM_3}, {"leftbracket", VK_OEM_4},
    {"backslash", VK_OEM_5}, {"rightbracket", VK_OEM_6}, {"quote", VK_OEM_7},
    {"f1", VK_F1}, {"f2", VK_F2}, {"f3", VK_F3}, {"f4", VK_F4},
    {"f5", VK_F5}, {"f6", VK_F6}, {"f7", VK_F7}, {"f8", VK_F8},
    {"f9", VK_F9}, {"f10", VK_F10}, {"f11", VK_F11}, {"f12", VK_F12},
    {"f13", VK_F13}, {"f14", VK_F14}, {"f15", VK_F15}, {"f16", VK_F16},
    {"f17", VK_F17}, {"f18", VK_F18}, {"f19", VK_F19}, {"f20", VK_F20},
};

// "cmd" is the Mac name; on Windows the equivalent physical key is Win.
const std::map<std::string, WORD> kMods = {
    {"ctrl", VK_CONTROL},
    {"alt", VK_MENU},
    {"shift", VK_SHIFT},
    {"cmd", VK_LWIN},
    {"lctrl", VK_LCONTROL}, {"rctrl", VK_RCONTROL},
    {"lalt", VK_LMENU}, {"ralt", VK_RMENU},
    {"lshift", VK_LSHIFT}, {"rshift", VK_RSHIFT},
    {"lcmd", VK_LWIN}, {"rcmd", VK_RWIN},
};

// A virtual-key plus its extended-key bit. Numpad Enter and Return are the same
// VK_RETURN and are told apart only by that bit, so the flag has to travel with the key
// rather than be derived from the virtual-key alone.
struct KeyCode {
  WORD vk = 0;
  bool extended = false;

  explicit operator bool() const { return vk != 0; }
  bool operator==(const KeyCode& other) const {
    return vk == other.vk && extended == other.extended;
  }
};

struct Hold {
  std::string id;
  KeyCode key;  // vk == 0 when the combo is modifiers only
  std::vector<KeyCode> mods;
};

// Everything currently held, in press order, so we can always release it — even if the
// plugin dies mid-hold. A stuck key is the worst failure here.
std::vector<Hold> g_holds;

// Is this key down on behalf of any hold? Keys are never pressed twice or released while
// another hold still wants them.
bool IsDown(const KeyCode& key) {
  for (const Hold& hold : g_holds) {
    if (hold.key == key) return true;
    if (std::find(hold.mods.begin(), hold.mods.end(), key) != hold.mods.end()) return true;
  }
  return false;
}

bool IsExtendedKey(WORD vk) {
  switch (vk) {
    case VK_RCONTROL:
    case VK_RMENU:
    case VK_LWIN:
    case VK_RWIN:
    case VK_INSERT:
    case VK_DELETE:
    case VK_HOME:
    case VK_END:
    case VK_PRIOR:
    case VK_NEXT:
    case VK_LEFT:
    case VK_RIGHT:
    case VK_UP:
    case VK_DOWN:
      return true;
    default:
      return false;
  }
}

// Names whose extended bit cannot be recovered from the virtual-key, because another
// name maps to the same one.
KeyCode Resolve(const std::string& name, WORD vk) {
  return KeyCode{vk, IsExtendedKey(vk) || name == "numpadenter"};
}

void SendKey(const KeyCode& key, bool down) {
  INPUT input = {};
  input.type = INPUT_KEYBOARD;
  input.ki.wVk = key.vk;
  input.ki.dwFlags = (down ? 0 : KEYEVENTF_KEYUP) |
                     (key.extended ? KEYEVENTF_EXTENDEDKEY : 0);
  SendInput(1, &input, sizeof(INPUT));
}

void ReleaseHold(const std::string& id) {
  auto it = std::find_if(g_holds.begin(), g_holds.end(),
                         [&](const Hold& hold) { return hold.id == id; });
  if (it == g_holds.end()) return;

  const Hold hold = *it;
  g_holds.erase(it);

  // Reverse order: key first, then modifiers — but only what no other hold still wants.
  if (hold.key && !IsDown(hold.key)) SendKey(hold.key, false);
  for (auto m = hold.mods.rbegin(); m != hold.mods.rend(); ++m) {
    if (!IsDown(*m)) SendKey(*m, false);
  }
}

void ReleaseAll() {
  while (!g_holds.empty()) {
    ReleaseHold(g_holds.back().id);
  }
}

void Press(const std::string& id, const std::vector<KeyCode>& mods, const KeyCode& key) {
  ReleaseHold(id);
  for (const KeyCode& m : mods) {
    if (!IsDown(m)) SendKey(m, true);
  }
  if (key && !IsDown(key)) SendKey(key, true);
  g_holds.push_back(Hold{id, key, mods});
}

// Tap a combo *without* disturbing any hold — this is what makes a "before release"
// hotkey different from an "after release" one. Modifiers already down are reused rather
// than re-pressed, and a tap key identical to a held key is skipped: its key-up would
// cancel the hold that owns it.
void TapOver(const std::vector<KeyCode>& mods, const KeyCode& key) {
  std::vector<KeyCode> extra;
  for (const KeyCode& m : mods) {
    if (!IsDown(m)) extra.push_back(m);
  }

  for (const KeyCode& m : extra) SendKey(m, true);
  if (key && !IsDown(key)) {
    SendKey(key, true);
    SendKey(key, false);
  } else if (key) {
    std::cerr << "keyholder: skipping tap of the key already held\n";
  }
  for (auto it = extra.rbegin(); it != extra.rend(); ++it) SendKey(*it, false);
}

BOOL WINAPI OnConsoleEvent(DWORD) {
  ReleaseAll();
  ExitProcess(0);
  return TRUE;
}

std::string Lower(std::string s) {
  std::transform(s.begin(), s.end(), s.begin(),
                 [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
  return s;
}

}  // namespace

int main() {
  SetConsoleCtrlHandler(OnConsoleEvent, TRUE);

  std::string line;
  while (std::getline(std::cin, line)) {
    std::istringstream stream(line);
    std::string verb;
    stream >> verb;

    if (verb == "U") {
      std::string id;
      if (!(stream >> id)) {
        std::cerr << "keyholder: bad command: " << line << "\n";
        continue;
      }
      ReleaseHold(id);
      continue;
    }
    if (verb != "D" && verb != "T") continue;

    std::string id, mod_list, key_name;
    const bool is_hold = verb == "D";
    if (is_hold && !(stream >> id)) {
      std::cerr << "keyholder: bad command: " << line << "\n";
      continue;
    }
    if (!(stream >> mod_list >> key_name)) {
      std::cerr << "keyholder: bad command: " << line << "\n";
      continue;
    }

    KeyCode key_code;
    const std::string key_lower = Lower(key_name);
    auto key = kKeys.find(key_lower);
    if (key_name != "-" && key == kKeys.end()) {
      std::cerr << "keyholder: unknown key: " << key_name << "\n";
      continue;
    }
    if (key != kKeys.end()) key_code = Resolve(key_lower, key->second);

    std::vector<KeyCode> mods;
    if (mod_list != "-") {
      std::istringstream mod_stream(mod_list);
      std::string name;
      while (std::getline(mod_stream, name, ',')) {
        const std::string mod_lower = Lower(name);
        auto mod = kMods.find(mod_lower);
        if (mod != kMods.end()) mods.push_back(Resolve(mod_lower, mod->second));
      }
    }

    if (mods.empty() && !key_code) continue;
    if (is_hold) {
      Press(id, mods, key_code);
    } else {
      TapOver(mods, key_code);
    }
  }

  // stdin closed — Stream Deck killed the plugin. Never leave a key down.
  ReleaseAll();
  return 0;
}
